platform-sdk/swirlds-common/src/test/java/com/swirlds/common/threading/SyncPermitProviderTest.java

```diff
@@ -0,0 +1,142 @@
+
+    /**
+     * Verify that once all permits are acquired, further attempts to acquire fail.
+     */
+    @Test
+    void testAllPermitsAcquired() {
+        final int numPermits = 9;
+        final SyncPermitProvider syncPermitProvider = new SyncPermitProvider(numPermits);
+
+        assertEquals(numPermits, syncPermitProvider.getNumAvailable(), "all permits should be available");
+
+        final List<MaybeLocked> permits = new ArrayList<>(numPermits);
+
+        // Acquire all the permits
+        for (int i = 0; i < numPermits; i++) {
+            final MaybeLocked maybeLocked = syncPermitProvider.tryAcquire();
+            permits.add(maybeLocked);
+            assertTrue(maybeLocked.isLockAcquired(), "first acquire should succeed");
+            assertEquals(
+                    numPermits - i - 1,
+                    syncPermitProvider.getNumAvailable(),
+                    "one less permit should be available when a permit is acquired");
+        }
+
+        // Attempts to acquire more permits should fail
+        final MaybeLocked shouldNotAcquire = syncPermitProvider.tryAcquire();
+        assertFalse(shouldNotAcquire.isLockAcquired(), "no further permits should be able to be acquired");
+
+        // Releasing permits should result in more permits being available
+        for (int i = 0; i < numPermits; i++) {
+            final MaybeLocked maybeLocked = permits.get(i);
+            maybeLocked.close();
+            assertEquals(
+                    i + 1,
+                    syncPermitProvider.getNumAvailable(),
+                    "one more permit should be available when a permit is released");
+        }
+    }
+
+    @Test
+    void testWaitForAllSyncsToFinish() {
+        final int numPermits = 3;
+        final SyncPermitProvider syncPermitProvider = new SyncPermitProvider(numPermits);
+
+        final List<MaybeLocked> permits = new ArrayList<>(numPermits);
+        // Acquire all the permits
+        for (int i = 0; i < numPermits; i++) {
+            final MaybeLocked maybeLocked = syncPermitProvider.tryAcquire();
+            permits.add(maybeLocked);
+            assertTrue(maybeLocked.isLockAcquired());
+        }
+
+        // Attempts to acquire more permits should fail
+        final MaybeLocked shouldNotAcquire = syncPermitProvider.tryAcquire();
+        assertFalse(shouldNotAcquire.isLockAcquired(), "no further permits should be able to be acquired");
+
+        final AtomicBoolean waitComplete = new AtomicBoolean(false);
+
+        // Have a separate thread wait for syncs to finish
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+        final Future<Void> future = executorService.submit(() -> {
+            syncPermitProvider.waitForAllSyncsToFinish();
+            waitComplete.set(true);
+            return null;
+        });
+
+        try {
+            // wait a bit, to give waitForAllSyncsToFinish time to potentially misbehave
+            MILLISECONDS.sleep(50);
```

Comment:
Cody created some nice assert utils for waiting up to a set amount of time for a condition to be true. Check out `com.swirlds.common.test.AssertionUtils.assertEventuallyTrue`